### PR TITLE
Shutdown TCP ServerSocket when parent process ends

### PR DESCRIPTION
**Pull request by andfoy**
This PR enables to interrupt actual TCP server when parent server ends, as the current behaviour only closes the communication channel

### CODE REVIEW

File: pyls/_utils.py [hunk 1]
```diff
@@ -146,6 +147,15 @@ def clip_column(column, lines, line_number):
     return min(column, max_column)
 
 
+def interrupt_process(pid=None):
```
Comment (gatesn) [inline pyls/_utils.py]: Why are we sending a sigint to ourselves? Python’s TCP server polls for shutdown by default every 500ms - can’t we just call shutdown on the server?

https://docs.python.org/2/library/socketserver.html#SocketServer.BaseServer.serve_forever

File: pyls/python_ls.py [hunk 2]
```diff
@@ -120,6 +123,9 @@ def m_exit(self, **_kwargs):
         self._jsonrpc_stream_reader.close()
         self._jsonrpc_stream_writer.close()
 
+        if self._is_tcp and self._check_parent_process:
```
Comment (gatesn) [inline pyls/python_ls.py]: I don’t like forking the logic here. If there’s a need to send a sigint to ourselves then why not do it for both stdio and tcp?

File: pyls/_utils.py (hunk 1 above)
Comment (andfoy) [reply pyls/_utils.py]: For some reason, the interrupt is not working, I'll take a look more carefully, I'm aware this is not the most appropriate way of handling this

File: pyls/python_ls.py (hunk 2 above)
Comment (andfoy) [reply pyls/python_ls.py]: See above.

**Comment by ccordoba12** [comment]
@gatesn, I think the last commit by @andfoy implements a solution that meets your requirements. Do you agree with it?

File: pyls/python_ls.py [hunk 3]
```diff
@@ -187,19 +198,19 @@ def m_initialize(self, processId=None, rootUri=None, rootPath=None, initializati
...
                     log.info("parent process %s is not alive", pid)
                     self.m_exit()
-                log.debug("parent process %s is still alive", pid)
-                threading.Timer(PARENT_PROCESS_WATCH_INTERVAL, watch_parent_process, args=[pid]).start()
-
-            watching_thread = threading.Thread(target=watch_parent_process, args=(processId,))
-            watching_thread.daemon = True
-            watching_thread.start()
+                else:
+                    log.debug("parent process %s is still alive", pid)
```
Comment (ccordoba12) [inline pyls/python_ls.py]: Please remove this `log` entry to not fill the PyLS log with useless info.

**Review by ccordoba12 — COMMENTED** [review]
@andfoy, i left you a small comment, then this should be ready.

**Comment by ccordoba12** [comment]
@gatesn, we really need this in Spyder, so if I don't hear from you until Tuesday I'll merge it. Hope that's ok for you.

File: test/test_language_server.py [hunk 4]
```diff
@@ -72,12 +76,17 @@ def test_initialize(client_server):  # pylint: disable=redefined-outer-name
 @unix_only
```
Comment (ccordoba12) [inline test/test_language_server.py:76]: @andfoy, could we remove this decorator to see if things as expected on Windows? I mean, I think after PR #445 they should.

Comment (andfoy) [reply test/test_language_server.py:76]: The problem is that tests fail when we try to use multiprocessing on Windows

Comment (ccordoba12) [reply test/test_language_server.py:76]: ok, no problem then :+1:

**Review by ccordoba12 — APPROVED** [review]
Thanks @andfoy!

**Pull request merged**